framework/src/engine/legacy/legacy_chain_handler.ts

```diff
@@ -12,56 +12,202 @@
+		}
+
+		this._network.applyNodeInfo({
+			legacy: this._legacyConfig.brackets.map(bracket =>
+				Buffer.from(bracket.snapshotBlockID, 'hex'),
+			),
+		});
+
+		clearTimeout(this._syncTimeout);
+	}
+
+	private async _trySyncBlocks(bracket: LegacyBlockBracket, lastBlock: LegacyBlock) {
+		try {
+			await this.syncBlocks(bracket, lastBlock);
+		} catch (err) {
+			if (err instanceof PeerNotFoundWithLegacyInfo) {
+				// eslint-disable-next-line @typescript-eslint/no-misused-promises
+				this._syncTimeout = setTimeout(async () => {
+					await this._trySyncBlocks(bracket, lastBlock);
+				}, 120000); // 2 mints = (60 * 2) * 1000
+			} else {
+				throw err;
+			}
+		}
+	}
+
+	/**
+	 * Flow of syncing legacy blocks
+	 *
+	 * Check if we have `sync` property `true` in configuration
+	 * getConnectedPeers from network
+	 * Filter peers having their `legacy` buffer array property contains `snapshotBlockID`
+	 * If there is no peer, throw error, this error will be used in outside function to retry calling `syncBlocks` after x amount of time
+	 * Get a random peer from list of filtered peers with legacy info
+	 * Make a request to that random peer by calling its `getLegacyBlocksFromId` method with `data` property set to `legacyBlock.header.id`
+	 * Try to decode response data buffer into Block, apply penalty in case of error
+	 * Start saving parsed blocks one by one
+	 * repeat, if last parsed block height is still higher than bracket.startHeight
+	 * finally, update node info
+	 */
+	// eslint-disable-next-line @typescript-eslint/member-ordering
+	public async syncBlocks(bracket: LegacyBlockBracket, legacyBlock: LegacyBlock): Promise<void> {
+		const connectedPeers = (this._network.getConnectedPeers() as unknown) as Peer[];
+		const peersWithLegacyInfo = connectedPeers.filter(
+			peer =>
+				!!(peer.options as { legacy: Buffer[] }).legacy.find(snapshotBlockID =>
+					snapshotBlockID.equals(Buffer.from(bracket.snapshotBlockID, 'hex')),
+				),
+		);
+		if (!peersWithLegacyInfo) {
+			throw new PeerNotFoundWithLegacyInfo('No peer found with legacy info.');
+		}
+
+		const randomPeerIndex = Math.trunc(Math.random() * peersWithLegacyInfo.length - 1);
+		const { peerId } = peersWithLegacyInfo[randomPeerIndex];
+
+		const p2PRequestPacket: P2PRequestPacket = {
+			procedure: 'getLegacyBlocksFromId',
+			data: legacyBlock.header.id,
+		};
+
+		const response = await this._network.requestFromPeer({ ...p2PRequestPacket, peerId });
+		if (!(response.data as []).length) {
+			return;
+		}
+
+		// `data` is expected to hold blocks in DESC order
+		const { data } = response;
+		let legacyBlocks: LegacyBlock[];
+		try {
+			// this part is needed to make sure `data` returns ONLY `{ blocks: Buffer[] }` & not any extra field(s)
+			const { blocks } = codec.decode<{ blocks: Buffer[] }>(
+				getBlocksFromIdResponseSchema,
+				data as Buffer,
+			);
+			this._applyValidation(blocks);
+			legacyBlocks = blocks.map(block => decodeBlock(block).block);
```

Comment:
i think here you  should check whether blocks.length >0 , else you should applyPenalty 100 and this._network.applyPenaltyOnPeer({ peerId, penalty: 100 and call syncBlocks instead of return